### PR TITLE
Test logging uses display names

### PR DESCRIPTION
**Pull request by britter**
Test logging now logs the display name if it is available. Furthermore
getDisplayName() and getClassDisplayName() are moved from
TestDescriptorInternal to TestDescriptor making them available in
beforeTest and afterTest callbacks.

Fixes: #10983 

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [x] Recognize contributor in release notes

### CODE REVIEW

**Comment by britter** [comment]
This needs discussion. The problem being one the edge case that somebody could set the display name to the name of the class to remove the package name from the log output. This does not seem to be possible because the JUnit `TestIdentifier` does not have information about where the display name was coming form.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultTestClassDescriptor.java [hunk 1]
```diff
@@ -35,12 +35,26 @@ public String getClassName() {
 
     @Override
     public String getDisplayName() {
-        return classDisplayName;
+        if (isInnerClass()) {
```
Comment (britter) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultTestClassDescriptor.java]: This if can be dropped

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java [hunk 2]
```diff
@@ -41,6 +42,21 @@
...
+     */
+    @Incubating
+    String getDisplayName();
+
+    /**
+     * The class name for display. It may be the same as or different from {@link #getClassName()}
+     * @return the class name for display.
+     */
+    @Incubating
+    String getClassDisplayName();
```
Comment (britter) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java]: Should we add tests that make sure this works in `beforeTest` and `afterTest` callbacks?

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/logging/AbstractTestLogger.java [hunk 3]
```diff
@@ -70,9 +70,9 @@ private String getEventPath(TestDescriptor descriptor) {
                 // but we nevertheless want to see the class name. We use "." rather than
                 // " > " as a separator to make it clear that the class is not a separate
                 // level. This matters when configuring granularity.
-                names.add(current.getClassName() + "." + current.getName());
```
Comment (britter) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/logging/AbstractTestLogger.java]: Don't change this

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultTestClassDescriptor.java [hunk 4]
```diff
@@ -35,12 +35,26 @@ public String getClassName() {
 
     @Override
     public String getDisplayName() {
```
Comment (britter) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultTestClassDescriptor.java]: Return the display name instead of having a separate logic here.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestDescriptorInternal.java [hunk 5]
```diff
@@ -36,18 +36,7 @@
     @Nullable
     Object getOwnerBuildOperationId();
 
-    /**
-     * The name for display. It may be the same as or different from {@link #getName()}
-     *
-     * @return the name for display.
-     */
-    String getDisplayName();
```
Comment (britter) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestDescriptorInternal.java]: I'm thinking about moving these changes to a separate PR so we can better promote it in the release notes. WDYT?

Comment (marcphilipp) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestDescriptorInternal.java]: Makes sense. 👍

Comment (bjliu) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestDescriptorInternal.java]: This is the change to move getDisplayName and getClassDisplayName from an internal class to the TestDescriptor? It does seem like a separate issue.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/logging/AbstractTestLogger.java [hunk 6]
```diff
@@ -72,7 +72,7 @@ private String getEventPath(TestDescriptor descriptor) {
                 // level. This matters when configuring granularity.
                 names.add(current.getClassName() + "." + current.getName());
             } else {
-                names.add(current.getName());
+                names.add(current.getDisplayName());
```
Comment (bjliu) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/logging/AbstractTestLogger.java]: When I was looking at this file, I was wondering: does this change work if there is no display name?

Comment (britter) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/logging/AbstractTestLogger.java]: The `getDisplayName()` was not annotated with `@Nullable` so I assumed it is always present.

**Comment by britter** [comment]
@marcphilipp please have a lost look at the recently changed test cases. I think we're good to go now.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java [hunk 7]
```diff
@@ -41,6 +42,15 @@
     @Nullable
     String getClassName();
 
+    /**
+     * The name for display. It may be the same as or different from {@link #getName()}
+     *
+     * @return the name for display.
```
Comment (marcphilipp) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java]: :o: The wording is a little inconsistent with the rest of the class.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitConsoleLoggingIntegrationTest.groovy [hunk 8]
```diff
@@ -41,7 +41,7 @@ class JUnitConsoleLoggingIntegrationTest extends JUnitMultiVersionIntegrationSpe
 
         then:
         outputContains("""
-org.gradle.JUnit4Test > badTest FAILED
+JUnit4Test > badTest FAILED
```
Comment (marcphilipp) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitConsoleLoggingIntegrationTest.groovy]: ❓ Why are these no longer present when executing the test with JUnit 4?

Comment (britter) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitConsoleLoggingIntegrationTest.groovy]: This is a multi version spec that executes tests for JUnit 4.12 and JUnit Platform. In case JUnit 4.12 test are executed the package prefix is still present. I will change the test to reflect this.

Comment (marcphilipp) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitConsoleLoggingIntegrationTest.groovy]: Never mind, they are, we just don't test for them anymore.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java (hunk 7 above)
Comment (britter) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java]: Fixed
